apache2/apache2_util.c

```diff
@@ -465,28 +465,10 @@ static void internal_log_ex(request_rec *r, directory_config *dcfg, modsec_rec *
         else requestheaderhostname = "";
 
 #ifdef WAF_JSON_LOGGING_ENABLE
-        if (msc_waf_log_reopened) {
-            rc = apr_file_open(&msc_waf_log_fd, msc_waf_log_path,
-                   APR_WRITE | APR_APPEND | APR_CREATE | APR_BINARY,
-                   CREATEMODE | APR_WREAD, msc_waf_log_cmd->pool);
-
-            if (rc != APR_SUCCESS) {
-                #if AP_SERVER_MAJORVERSION_NUMBER > 1 && AP_SERVER_MINORVERSION_NUMBER > 2
-                	ap_log_rerror(APLOG_MARK, APLOG_ERR | APLOG_NOERRNO, 0, r,
-                            "ModSecurity not able to reopen %s file", WAF_LOG_UTIL_FILE);
-                #else
-                        ap_log_error(APLOG_MARK, APLOG_ERR | APLOG_NOERRNO, 0, r->server,
-                            "ModSecurity not able to reopen %s file", WAF_LOG_UTIL_FILE);
-                #endif
-            }
-
-            msc_waf_log_reopened = 0;
-        }
-
         const char* scope = apr_table_get(r->notes, WAF_POLICY_SCOPE);
         const char* scope_name = apr_table_get(r->notes, WAF_POLICY_SCOPE_NAME);
 
-        send_waf_log(msr->modsecurity->wafjsonlog_lock, msc_waf_log_fd, str1, r->useragent_ip ? r->useragent_ip : r->connection->client_ip, log_escape(msr->mp, r->uri), (!msr->allow_scope) ? dcfg->is_enabled : msr->allow_scope, r->hostname, r->log_id, r, dcfg->waf_policy_id, scope ? scope : "", scope_name ? scope_name : "");
+        send_waf_log(msr->modsecurity->wafjsonlog_lock, &msc_waf_log_fd, str1, r->useragent_ip ? r->useragent_ip : r->connection->client_ip, log_escape(msr->mp, r->uri), (!msr->allow_scope) ? dcfg->is_enabled : msr->allow_scope, r->hostname, r->log_id, r, dcfg->waf_policy_id, scope ? scope : "", scope_name ? scope_name : "");
```

Comment:
could a race condition happen here, where &msc_waf_log_fd evaluates to some addr, then this thread gets preempted, then modsecReopenLogfileIfNeeded runs?

Reply:
I think the address of the file descriptor should never change for the lifetime of the process as we are merely assigning values to it. That was precisely the reason why I suggested that we pass the pointer to the file descriptor instead of simply passing it by value. This way, we make sure if the content of the variable changes, we take the right one.